src/States.cpp

```diff
@@ -0,0 +1,404 @@
+  }
+
+  std::vector<ale::Vec3d> States::getPositions() const {
+    // extract positions from state vector
+    std::vector<ale::Vec3d> positions; 
+    
+    for(ale::State state : m_states) {
+        positions.push_back(state.position);
+    }
+    return positions; 
+  }
+
+
+  std::vector<ale::Vec3d> States::getVelocities() const {
+    // extract velocities from state vector
+    std::vector<ale::Vec3d> velocities; 
+    
+    for(ale::State state : m_states) {
+        velocities.push_back(state.velocity);
+    }
+    return velocities; 
+  }
+
+
+  std::vector<double> States::getTimes() const {
+    return m_ephemTimes; 
+  }
+
+
+  int States::getReferenceFrame() const {
+    return m_refFrame; 
+  }
+
+
+  bool States::hasMinimizedCache() const {
+    return m_minimizedCache;
+  }
+
+
+  bool States::hasVelocity() const {
+    std::vector<ale::Vec3d> velocities = getVelocities(); 
+    bool allZero = std::all_of(velocities.begin(), velocities.end(), [](ale::Vec3d vec) 
+                               { return vec.x==0.0 && vec.y==0.0 && vec.z==0.0; });
+    return !allZero;
+  }
+
+
+  ale::State States::getState(double time, ale::interpolation interp) const {
+    std::vector<double> xs, ys, zs, vxs, vys, vzs; 
+    for (ale::State state : m_states) {
+      xs.push_back(state.position.x);
+      ys.push_back(state.position.y);
+      zs.push_back(state.position.z);
+      vxs.push_back(state.velocity.x);
+      vys.push_back(state.velocity.y);
+      vzs.push_back(state.velocity.z);
+    }
+
+    ale::Vec3d position, velocity; 
+
+    if (!hasMinimizedCache()) {
+      position = {interpolateState(xs,  m_ephemTimes, time, interp, 0),
+                  interpolateState(ys,  m_ephemTimes, time, interp, 0),
+                  interpolateState(zs,  m_ephemTimes, time, interp, 0)};
+
+      velocity = {interpolateState(vxs, m_ephemTimes, time, interp, 0),
+                  interpolateState(vys, m_ephemTimes, time, interp, 0),
+                  interpolateState(vzs, m_ephemTimes, time, interp, 0)};
+    }
+    else {
+      // We have a minimized cache
+      double baseTime = (m_ephemTimes.at(0) + m_ephemTimes.at(m_ephemTimes.size() - 1)) / 2.;
+      double timeScale = 1.0;
+
+      std::vector<double> scaledEphemTimes; 
+      for(unsigned int i = 0; i < m_ephemTimes.size(); i++) {
+        scaledEphemTimes.push_back((m_ephemTimes[i] - baseTime) / timeScale);
+      }
+
+      double sTime = (time - baseTime) / timeScale;
+      position.x = ale::evaluateCubicHermite(sTime, vxs, scaledEphemTimes, xs);
+      position.y = ale::evaluateCubicHermite(sTime, vys, scaledEphemTimes, ys);
+      position.z = ale::evaluateCubicHermite(sTime, vzs, scaledEphemTimes, zs);
+
+      velocity.x = ale::evaluateCubicHermiteFirstDeriv(sTime, vxs, scaledEphemTimes, xs);
+      velocity.y = ale::evaluateCubicHermiteFirstDeriv(sTime, vys, scaledEphemTimes, ys);
+      velocity.z = ale::evaluateCubicHermiteFirstDeriv(sTime, vzs, scaledEphemTimes, zs);
```

Comment:
See my comments above about changing these functions to operate on state objects.